File: leetcode/101-200/leetcode148.cpp
#include <iostream>
using namespace std;
struct ListNode {
    int val;
    ListNode* next;
    ListNode()
        : val(0)
        , next(nullptr)
    {
    }
    ListNode(int x)
        : val(x)
        , next(nullptr)
    {
    }
    ListNode(int x, ListNode* next)
        : val(x)
        , next(next)
    {
    }
};

// class Solution {
// public:
//     ListNode* sortList(ListNode* head)
//     {
//         if (!head || !head->next)
//             return head;
//         ListNode *sortedHead, *sortedPos, *sortedNext;
//         sortedHead = sortedPos = new ListNode(0);
//         while (head) {
//             sortedPos = sortedHead;
//             bool checker = false;
//             while (sortedPos->next) {
//                 if (sortedPos->next->val >= head->val) {
//                     sortedNext = sortedPos->next;
//                     sortedPos->next = head;
//                     head = head->next;
//                     sortedPos->next->next = sortedNext;
//                     checker = true;
//                     break;
//                 }
//                 sortedPos = sortedPos->next;
//             }
//             if (!checker) {
//                 sortedPos->next = head;
//                 head = head->next;
//                 sortedPos->next->next = nullptr;
//             }
//         }
//         return sortedHead->next;
//     }
// };
class Solution {
public:
    ListNode* sortList(ListNode* head)
    {
        if (!head || !head->next)
            return head;
        ListNode *fast, *slow, *prev;
        fast = slow = head;
        prev = nullptr;
        while (fast && fast->next) {
            prev = slow;
            slow = slow->next;
            fast = fast->next->next;
        }

        prev->next = nullptr;
        ListNode* left = sortList(head);
        ListNode* right = sortList(slow);
        return mergeSort(left, right);
    }

    ListNode* mergeSort(ListNode* l1, ListNode* l2)
    {
        ListNode *neoHead, *pos;
        neoHead = pos = new ListNode(0);
        while (l1 && l2) {
            if (l1->val <= l2->val) {
                pos->next = l1;
                l1 = l1->next;
            } else {
                pos->next = l2;
                l2 = l2->next;
            }
            pos = pos->next;
        }
        if (l1)
            pos->next = l1;
        if (l2)
            pos->next = l2;

        pos = neoHead->next;
        delete neoHead;
        return pos;
    }
};
int main(void) { return 0; }

File: leetcode/100-999/leetcode148.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    def __repr__(self) -> str:
        cur = self
        res = ""
        while cur:
            res += str(cur.val) + " "
            cur = cur.next
        return res

from typing import Optional


class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return Solution.chaewon(head)

    @staticmethod
    def chaewon(head: Optional[ListNode]) -> Optional[ListNode]:
        # split
        if not head or not head.next:
            return head

        left, right = Solution.split(head)

        left = Solution.chaewon(left)
        right = Solution.chaewon(right)

        sorted_list = Solution.merge(left, right)
        
        return sorted_list

    @staticmethod
    def split(a: ListNode) -> tuple:
        slow, fast = a, a.next
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next

        fast = slow.next
        slow.next = None
        return a, fast

    @staticmethod
    def merge(a: ListNode, b: ListNode) -> ListNode:
        res_head = cur = ListNode()

        while a and b:
            if a.val <= b.val:
                cur.next = a
                a = a.next
            else:
                cur.next = b
                b = b.next
            cur = cur.next
            cur.next = None

        if a:
            cur.next = a
        if b:
            cur.next = b

        return res_head.next


s = Solution()
head = ListNode(3)
head = ListNode(1, head)
head = ListNode(2, head)
head = ListNode(4, head)
head = s.sortList(head=head)
print(head)
